crates/starknet-types-core/src/short_string/mod.rs

```diff
@@ -0,0 +1,179 @@
+//! Cairo short string
+//!
+//! The cairo language make it possible to create `Felt` values at compile time from so-called "short string".
+//! See https://docs.starknet.io/archive/cairo-101/strings/ for more information and syntax.
+//!
+//! This modules allows to mirror this behaviour in Rust, by leveraging type safety.
+//! A `ShortString` is string that have been checked and is guaranted to be convertible into a valid `Felt`.
+//! It checks that the `String` only contains ascii characters and is no longer than 31 characters.
+//!
+//! The convesion to `Felt` is done by using the internal ascii short string as bytes and parse those as a big endian number.
+
+#[cfg(not(feature = "std"))]
+use crate::felt::alloc::string::{String, ToString};
+use crate::felt::Felt;
+
+/// A cairo short string
+///
+/// Allow for safe conversion of cairo short string `String` into `Felt`,
+/// as it is guaranted that the value it contains can be represented as a felt.
+#[repr(transparent)]
+#[derive(Debug, Clone, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ShortString(String);
+
+impl core::fmt::Display for ShortString {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl From<ShortString> for Felt {
+    fn from(ss: ShortString) -> Self {
+        let bytes = ss.0.as_bytes();
+
+        let mut buffer = [0u8; 32];
+        buffer[(32 - bytes.len())..].copy_from_slice(bytes);
```

Comment:
if `bytes.len() > 32` you're fucked

Reply:
Due to the way `ShortString` is initialized, see `impl TryFrom<String> for ShortString`, there is no way `bytes.len()` can be greater than 32.

The string is guaranteed to be `31` chars at most, and the `is_ascii` guarantee that each one of those characters is made of 1 byte. Therefore, `bytes.len()` cannot be more than `31`

Reply:
The following snippet:

```rust
let ko_string = String::from("This is a 33 characters string...");
let _ = Felt::from(ShortString(ko_string));
```

fails with this error:

```
thread [...] panicked at crates/starknet-types-core/src/short_string/mod.rs:36:16:
attempt to subtract with overflow
```

If the idea is to panic when the string is too long then I think it would be great to have a more descriptive error message

Reply:
Now that I realize, I tested it by adding a test case in the same file so I could create a `ShortString` using the inner value.
So I think that, instead of improving the error message (because that from function would never panic), it would be good to have a comment that explains why the subtraction would never underflow.

Reply:
@gabrielbosio yup the inner value is only accessible from this mod. Outside of it your only way to create one is to go through this impl `impl TryFrom<&str> for ShortString`

> why the subtraction would never underflow
I can add a comment, giving some context for sure